docs/extensions.rst

```diff
@@ -64,62 +230,121 @@ The following example illustrates a simple extension implementation:
-    from ${namespace_pkg}.awesome import awesome
+    from {namespace_pkg}.awesome import awesome
 
     def test_awesome():
         assert awesome() == "Awesome!"
     """
 
-    def extend_pyscaffold(scaffold):
-        """Define extension behavior."""
-        opts = scaffold.options
-
-        # PyScaffold can run arbitrary functions before and after generating
-        # the files.
-        scaffold.before_generate.append(lambda _: print("Awesome Start"))
-        scaffold.after_generate.append(lambda _: print("Awesome End"))
-
-        # Extra files can be added to the PyScaffold structure.
-        scaffold.merge_structure({
-            opts['project']: {
-                opts['package']: {
-                    'awesome.py': MY_AWESOME_FILE.format(opts)
-                    # When a leaf is a string, the content is written in the
-                    # file path indicated by the dictionary keys.
-                },
-                'tests': {
-                    'awesome_test.py': (
-                        MY_AWESOME_TEST.format(opts),
-                        scaffold.NO_OVERWRITE
-                    )
-                    # When a leaf is a tuple, the first element is used as
-                    # content for the file, while the second element is used
-                    # as the update rule for existing projects.
+    def extend_scaffold(actions, helpers):
+
+        # ...
+
+        def define_awesome_files(structure, opts):
+            structure = helpers.merge(structure, {
+                opts['project']: {
+                    opts['package']: {
+                        'awesome.py': MY_AWESOME_FILE.format(opts)
+                    },
+                    'tests': {
+                        'awesome_test.py': (
+                            MY_AWESOME_TEST.format(opts),
+                            helpers.NO_OVERWRITE
+                        )
+                    }
                 }
-            }
-        })
-
-        # Files can be directly added to the `structure` dict.
-        scaffold.structure['.python-version'] = ('3.6.1', scaffold.NO_OVERWRITE)
-
-        # The `ensure_file` method can be also used.
-        for filename in opts['awesome_files']:
-            scaffold.ensure_file(filename, content='AWESOME!',
-                                 update_rule=scaffold.NO_CREATE
-                                 path=[opts['project'], 'awesome'])
-
-        # The `reject_file` can be used to avoid files being generated.
-        del scaffold.reject_file('skeleton.py',
-                                 path=[opts['project'], opts['package']])
-
-Note that both :attr:`~pyscaffold.api.Scaffold.before_generate` and
-:attr:`~pyscaffold.api.Scaffold.after_generate` hooks also should be
-defined as a function of a single argument, a
-:class:`~pyscffold.api.Scaffold` instance.
+            })
+
+            structure['.python-version'] = ('3.6.1', helpers.NO_OVERWRITE)
+
+            for filename in opts['awesome_files']:
```

Comment:
Cool examples but at that point it might not be clear to the reader how to get `awesome_files` entry into the `opts` dictionary. One would need to do this somehow in a function that was registered before and adds `awesome_files` to the options, right?